builder/builder.go

```diff
@@ -0,0 +1,192 @@
+	relay        IRelay
+
+	builderSecretKey     *bls.SecretKey
+	builderPublicKey     boostTypes.PublicKey
+	builderSigningDomain boostTypes.Domain
+}
+
+func NewBuilder(sk *bls.SecretKey, bc IBeaconClient, relay IRelay, builderSigningDomain boostTypes.Domain) *Builder {
+	pkBytes := bls.PublicKeyFromSecretKey(sk).Compress()
+	pk := boostTypes.PublicKey{}
+	pk.FromSlice(pkBytes)
+
+	_, err := bc.onForkchoiceUpdate()
+	if err != nil {
+		log.Error("could not initialize beacon client", "err", err)
+	}
+
+	return &Builder{
+		beaconClient:     bc,
+		relay:            relay,
+		builderSecretKey: sk,
+		builderPublicKey: pk,
+
+		builderSigningDomain: builderSigningDomain,
+	}
+}
+
+func (b *Builder) onForkchoice(payloadAttributes *beacon.PayloadAttributesV1) {
+	dataJson, err := json.Marshal(payloadAttributes)
+	if err == nil {
+		log.Info("FCU", "data", string(dataJson))
+	}
+
+	nextSlot, err := b.beaconClient.onForkchoiceUpdate()
+	if err != nil {
+		return
+	}
+
+	if payloadAttributes != nil {
+		payloadAttributes.Slot = nextSlot
+		if vd, err := b.relay.GetValidatorForSlot(nextSlot); err == nil {
+			payloadAttributes.SuggestedFeeRecipient = [20]byte(vd.FeeRecipient)
+			payloadAttributes.GasLimit = vd.GasLimit
+		}
+	}
+}
+
+func (b *Builder) newSealedBlock(data *beacon.ExecutableDataV1, block *types.Block, payloadAttributes *beacon.PayloadAttributesV1) {
+	dataJson, err := json.Marshal(data)
+	if err == nil {
+		log.Info("newSealedBlock", "data", string(dataJson))
+	}
+	payload := executableDataToExecutionPayload(data)
+
+	vd, err := b.relay.GetValidatorForSlot(payloadAttributes.Slot)
+	if err != nil {
+		log.Error("could not get validator while submitting block", "err", err, "slot", payloadAttributes.Slot)
+		return
+	}
+
+	pubkey, err := boostTypes.HexToPubkey(string(vd.Pubkey))
+	if err != nil {
+		log.Error("could not parse pubkey", "err", err, "pubkey", vd.Pubkey)
+		return
+	}
+
+	value := new(boostTypes.U256Str).FromBig(block.Profit)
+	blockBidMsg := BuilderSubmitBlockRequestMessage{
+		Slot:                 payloadAttributes.Slot,
+		ParentHash:           payload.ParentHash,
+		BlockHash:            payload.BlockHash,
+		BuilderPubkey:        b.builderPublicKey,
+		ProposerPubkey:       pubkey,
+		ProposerFeeRecipient: boostTypes.Address(payloadAttributes.SuggestedFeeRecipient),
+		Value:                *value,
+	}
+
+	/* signature, err := boostTypes.SignMessage(blockBidMsg, b.builderSigningDomain, b.builderSecretKey)
+	if err != nil {
+		log.Error("could not sign builder bid", "err", err)
+		return
+	} */
+
+	signature := boostTypes.Signature{}
+	blockSubmitReq := BuilderSubmitBlockRequest{
+		Signature:        signature,
+		Message:          blockBidMsg,
+		ExecutionPayload: *payload,
+	}
+
+	err = b.relay.SubmitBlock(&blockSubmitReq)
+	if err != nil {
+		log.Error("could not submit block", "err", err)
+		return
+	}
+}
+
+func payloadToPayloadHeader(p *boostTypes.ExecutionPayload) (*boostTypes.ExecutionPayloadHeader, error) {
```

Comment:
reuse https://github.com/flashbots/go-boost-utils/blob/main/types/builder.go#L253

Reply:
and yes, we can remove the nil check and error response there if you prefer :)